docs/_docs/01_getting-started/quick-start.md

```diff
@@ -91,6 +105,29 @@ sources that terragrunt supports.
 You can deploy this example by copy pasting it into a folder and running `terragrunt apply`.
 
 
+## A note about using modules from the registry
+
+The key design of Terragrunt is to act as a preprocessor to convert **shared modules** in the registry into a **root
+module**. In Terraform, modules can be loosely categorized into two types:
+
+* **Root Module**: A Terraform module that is designed for running `terraform init` and the other workflow commands
+  (`apply`, `plan`, etc). This is the entrypoint module for deploying your infrastructure. Root modules are identified
+  by the presence of key blocks that setup configuration about how Terraform behaves, like `backend` blocks (for
+  configuring state) and `provider` blocks (for configuring how Terraform interacts with the cloud APIs).
+* **Shared Module**: A Terraform module that is designed to be included in other Terraform modules through `module`
+  blocks. These modules are missing many of the key blocks that are required for running the workflow commands of
+  terraform.
```

Comment:
I think this misses some important nuance. See my comment here: https://github.com/gruntwork-io/terragrunt/issues/1774#issuecomment-904475528

Reply:
To me, the current docs read as if they are encouraging users to turn "shared modules" into "root modules," but isn't that the very problem we're trying to avoid in #1774? Isn't that behavior precisely what is causing some module maintainers headaches? In that sense, I worry these docs would indeed make things worse.

On the other hand, explaining that Terragrunt is designed to deploy only root modules, but that includes partially configured root modules, takes a little more work, but is more accurate, and may reduce the problems in #1774.

Reply:
> To me, the new docs in this PR read as if they are encouraging users to turn "shared modules" into "root modules," but isn't that the very problem we're trying to avoid in #1774? Isn't that behavior precisely what is causing some module maintainers headaches? In that sense, I worry these docs would indeed make things worse.
> On the other hand, explaining that Terragrunt is designed to deploy only root modules, but that includes partially configured root modules, takes a little more work, but is more accurate, and may reduce the problems in #1774.

Something about explaining the service module distinction here was bothering me, but I couldn't quite point out why. After sleeping on it, I think I figured out what was bothering me: this distinction actually puts the burden on module maintainers, and not us, Terragrunt.

Taking a step back, the real problem we are trying to solve here is that Terragrunt limitations are reported as bugs in the registry modules that the module maintainers need to fix, and not something that Terragrunt should address.

Currently, there is no clear cut technical reason why something is a service module from Terragrunt's perspective, like the definition Terraform has for root vs shared. There are technical limitations that prevent Terragrunt from deploying certain types of modules, but that doesn't mean implementing those would suddenly turn a module into a service module as we defined it. AFAIK, there are no service modules in the Terraform public registry. Even our GCP modules, which technically support direct deployment from Terragrunt, are not service modules under our definition.

Without clear technical definitions, it puts a lot of burden on module maintainers to make that distinction. The only distinction they can make is tagging their modules as such (service vs library). I worry that this burden would actually exacerbate the exact problem we are trying to solve, by sending the message that it's the responsibility of the module maintainer to support Terragrunt, and the maintainer is somehow explicitly making the choice not to support Terragrunt when it doesn't work.

That is the exact mentality I want to shift here. I want to push the blame on to Terragrunt, not module maintainers.

Given that, by saying that Terragrunt is trying to convert shared modules on the registry into root modules, we at least make it clear that when it doesn't work, it's Terragrunt's fault, NOT the module maintainers.

Reply:
I agree that we don't have a clear technical definition of what Terragrunt is designed for, and the result is that people either hit a Terragrunt limitation, or try to use it for the wrong thing, and then blame the module maintainer, which we absolutely don't want. So I feel that it's important to do the following:

1. Provide a clear technical definition of what Terragrunt is designed to do. IMO, the "Terragrunt is a canonical Terraform wrapper script for deploying partially configured root modules" as I explained in https://github.com/gruntwork-io/terragrunt/issues/1774#issuecomment-904475528 is the right thing to aim for here, rather than trying to tell folks that Terragrunt is designed to turn _any_ shared module into a root module, which it's not.
1. List out the known limitations of Terragrunt. E.g., Talk about limits of what sorts of partially configured root modules we can deploy, the need to set `type` on input vars, `run-all` limitations, etc. 
1. Explain what to do if you hit a new limitation. Namely, file a bug in this repo, rather than bothering module maintainers. We can even be more explicit here: if the module works with pure Terraform, but not with Terragrunt, that's Terragrunt's fault, and not the module maintainers.

WDYT?

Reply:
I still disagree strongly (namely that, whether we like it or not, Terragrunt IS a tool to convert any shared module to root module), but I also don't have the energy to continue this discussion and question myself whether it's on the verge of bike shedding, so I am willing to disagree and commit here.

Did my best to satisfy myself and hopefully your recommendations in (1), (2), and (3) with e304ad6

Reply:
We should probably chat live to make sure we're on the same page here. 

One note: these doc updates are in the quick start guide... But that level of detail doesn't feel particularly quick. Perhaps we should move this new content to a different doc, and update the quick start guide with a "Heads up, not all Registry modules can be deployed with Terragrunt, see `(link to new doc)` for details."

Reply:
Good point! Done in 647c546